### PR TITLE
server: init external storage before starting SQL

### LINKED ISSUE #120022: backup,distsql,server: backups fail with error `cannot create external storage before init`
It seems likely that we're starting the distsql server, and then it is being sent a processor to run, before external storage is initialized.

Jira issue: CRDB-36435

### PR DESCRIPTION
**Pull request by yuzefovich**
Once we start SQL server (namely DistSQL server within it), we must be able to serve distributed DistSQL plans like backups. Previously, we would initialize the external storage _after_ having starting the SQL server, so distributed backup plans could run into an error. We _might_ need to perform the initialization after having set the SQL Instance ID, so we do the initialization in-between SQL instance ID setting and starting the DistSQL and PG servers.

I also adjusted the comment that was introduced in c3b8bcd4db976e64561d4153bc0a108c28c2c52f and became somewhat stale.

Fixes: #120022.

Release note: None

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -2055,21 +2055,9 @@ func (s *topLevelServer) PreStart(ctx context.Context) error {
...
-		); err != nil {
-			return err
-		}
-	}
-
-	// Initialize the external storage builders configuration params now that the
-	// engines have been created. The object can be used to create ExternalStorage
-	// objects hereafter.
+	// Initialize the external storage builders before starting the SQL server
+	// in order to be able to server distributed plans (which we need to do once
```
Comment (stevendanna) [inline pkg/server/server.go:2059]:
```suggestion
	// in order to be able to serve distributed plans (which we need to do once
```

**Review by stevendanna — COMMENTED** [review]
I'm unsure we can do this but I may just need to read more code.  Namely, I thought we needed the SQL server started in the case of tenants so that we have a SQL instance ID to be able to serve nodelocal requests and I think we need sql started to serve userfile requests.

Ill read some more code and follow up.

Otherwise it seems like a sensible idea, thanks for taking a look.

**Review by yuzefovich — COMMENTED** [review]
I adjusted the code to do the initialization right before starting DistSQL and PG servers (which addresses the SQL instance ID concern), but you're saying that we might need to have SQL started already? I'm not familiar with the external storage and how it's used - is it because we might use the internal executor for some things? If so, it seems impossible to address #120022 :smiley: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123991#-:-NxoA9Fh5RAJcQJ4Z03a:bsf2stm)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @nkodali, and @stevendanna)

___
*[`pkg/server/server.go` line 2059 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/123991#gh-1597976427:-NxoA8bj53_2piVpf4zS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e9033494a54571b235f61d1edfc0f26739ad3ca/pkg/server/server.go#L2059)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

```suggestion
	// in order to be able to serve distributed plans (which we need to do once
```
</blockquote></details>

Done.

**Review by stevendanna — COMMENTED** [review]
> but you're saying that we might need to have SQL started already? I'm not familiar with the external storage and how it's used - is it because we might use the internal executor for some things?

Yes, external storage has 2 implementations,`userfile://` and `external://`, that use an internal executor. `userfile://` uses the internal executor reads and writes files to a table in SQL. `external://` allows the user to created aliases for storage locations. These alias are stored in SQL. 

> If so, it seems impossible to address # 120022

A few potential options:

1. We previously had code in the cloud storage layer that would block the caller until external storage was initialised. This was used by disagg storage and online restore, which need to have external storage access very early.  However, we removed it since it quickly exposed deadlocks and instead added "EarlyBootExternalStorageProviders" that can be used without initialisation. But, perhaps we bring the blocking mechanism back in a more limited way to handle this edge case.

2. Complete speculation with little knoweldge of DistSQL internals: Carve up the DistSQL startup, separating the ability to use DistSQL for flows initiated from this gateway from the ability to accept remote work.  I imagine this is a lot of work.

3. Do we know where in this function is the earliest we can safely use the internal executor? Is it right after the pgWire server is started?  It's not a _real_ fix, but moving this init to right after the pgWire startup still drastically reduces the scope of the race because we won't be waiting for all the migration code.

4. If we make sure that `userfile://` and `external://` don't  panic if they happen to get used before the ie is actually ready, then we could do what you have here. Userfile is not heavily used outside of tests and `external://` hasn't seen much uptake, so this would still be an improvement for most users.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123991#-:-NxpKeI2CH755jiW0nPT:b-15zlip)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt)

**Comment by yuzefovich** [comment]
Thanks for all the context. I think I'll close this PR since it's not that straightforward, and I don't have good enough grasp on all things to push this forward.

To me options 1 or 4 seem like the right way to go. Option 2 seems hard to achieve (once DistSQL server is started, we propagate information about that via gossip, so other nodes might be using recently started node for DistSQL work; we do want to remove this gossip usage in the future though). For option 3 I looked for a bit and I actually have no idea at which point during the server startup the internal executor is ready 😅, and it doesn't really solve the problem.

**Comment by stevendanna** [comment]
Your argument makes sense to me, I _think_ what I was worried about was accepting DistSQL flows that have been started on other nodes.  Although, now that DistSQL uses sqlinstances based planning, perhaps that is impossible. Or maybe it was always impossible and I was just confused.

**Pull request closed without merging**